src/fenics_constitutive/solver.py

```diff
@@ -0,0 +1,318 @@
+from __future__ import annotations
+
+import basix
+import dolfinx as df
+import numpy as np
+import ufl
+from petsc4py import PETSc
+
+from .interfaces import IncrSmallStrainModel
+from .maps import SubSpaceMap, build_subspace_map
+from .stress_strain import ufl_mandel_strain
+
+
+def build_history(
+    law: IncrSmallStrainModel, mesh: df.mesh.Mesh, q_degree: int
+) -> df.fem.Function | dict[str, df.fem.Function] | None:
+    """Build the history space and function(s) for the given law.
+
+    Args:
+        law: The constitutive law.
+        mesh: Either the full mesh for a homogenous domain or the submesh.
+        q_degree: The quadrature degree.
+
+    Returns:
+        The history function(s) for the given law.
+
+    """
+    match law.history_dim:
+        case int():
+            Qh = ufl.VectorElement(
+                "Quadrature",
+                mesh.ufl_cell(),
+                q_degree,
+                quad_scheme="default",
+                dim=law.history_dim,
+            )
+            history_space = df.fem.FunctionSpace(mesh, Qh)
+            history = df.fem.Function(history_space)
+        case None:
+            history = None
+        case dict():
+            history = {}
+            for key, value in law.history_dim.items():
+                if isinstance(value, int):
+                    Qh = ufl.VectorElement(
+                        "Quadrature",
+                        mesh.ufl_cell(),
+                        q_degree,
+                        quad_scheme="default",
+                        dim=value,
+                    )
+                elif isinstance(value, tuple):
+                    Qh = ufl.TensorElement(
+                        "Quadrature",
+                        mesh.ufl_cell(),
+                        q_degree,
+                        quad_scheme="default",
+                        shape=value,
+                    )
+                history_space = df.fem.FunctionSpace(mesh, Qh)
+                history[key] = df.fem.Function(history_space)
+    return history
+
+
+class IncrSmallStrainProblem(df.fem.petsc.NonlinearProblem):
+    def __init__(
+        self,
+        laws: list[tuple[IncrSmallStrainModel, np.ndarray]] | IncrSmallStrainModel,
+        u: df.fem.Function,
```

Comment:
`u` is the unknown that is solved for. Depending on how the loop over the load steps is written, this can either be the total displacement field $u$ or the displacement correction $\Delta u$. Just to keep in mind.

Reply:
Actually, `u` is always the full displacement field $u$. I changed the calcualtion of the strain/gradient of displacements to `nabla_grad(u-u0)`. Depending on the update of variables, the strain is just changed from total to incremental when we update `u0`, but `u` never represents $\Delta u$

For the future, we can replace the current `update` method with a more fine-grained `update_history` and `update_configuration` or something like that

Reply:
Ah okay, I missed the `nabla_grad(u - u0)` line.